e2e_context_test.go

```diff
@@ -0,0 +1,305 @@
+package httpexpect
+
+import (
+	"context"
+	"fmt"
+	"github.com/stretchr/testify/assert"
+	"net/http"
+	"net/http/httptest"
+	"strings"
+	"testing"
+	"time"
+)
+
+// the waitHandler sleeps and fails (500) for a number of calls and succeeds
+// (200) in the last one
+type waitHandler struct {
+	mux            *http.ServeMux
+	callIdx        int
+	callsDurations []time.Duration
+	callCount      int
```

Comment:
Minor concern: since callCount is accessed from two goroutines (from server and from test), it's better to make it atomic or guard it with a mutex; otherwise strictly speaking it's not guaranteed that the test will see the up-to-date value. Also, I'd hide reading callCount under a method.

Reply:
Addressed with d3445c9ae533afb0f81f3db5594fab5db5bc284a but overall the WaitXXXX handlers are not thread-safe.

Reply:
Thanks,

> but overall the WaitXXXX handlers are not thread-safe.

should be OK for tests since we're not issuing concurrent requests.